src/Chunk.ts

```diff
@@ -673,160 +688,244 @@ export default class Chunk {
+					if (namespace.renderFirst()) hoistedSource += '\n' + rendered;
+					else magicString.addSource(new MagicString(rendered));
+				}
+			}
+		}
 
-						if (options.sourcemap) {
-							timeStart('sourcemap', 3);
+		// resolve external module paths
+		for (let external of this.dependencies) {
+			if (!(external instanceof ExternalModule)) continue;
+			external.setRenderPath(options, this.entryModule && this.entryModule.id);
+		}
 
-							let file = options.file ? options.sourcemapFile || options.file : this.id;
-							if (file) file = resolve(typeof process !== 'undefined' ? process.cwd() : '', file);
+		if (hoistedSource) magicString.prepend(hoistedSource + '\n\n');
 
-							if (
-								this.graph.hasLoaders ||
-								this.graph.plugins.find(plugin =>
-									Boolean(plugin.transform || plugin.transformBundle)
-								)
-							) {
-								let decodedMap = magicString.generateDecodedMap({});
-								map = collapseSourcemaps(this, file, decodedMap, usedModules, bundleSourcemapChain);
-							} else {
-								map = magicString.generateMap({ file, includeContent: true });
-							}
+		this.renderedSource = magicString.trim();
 
-							map.sources = map.sources.map(normalize);
+		if (
+			this.getExportNames().length === 0 &&
+			this.getImportIds().length === 0 &&
+			this.renderedSource.isEmpty()
+		) {
+			this.graph.warn({
+				code: 'EMPTY_BUNDLE',
+				message: 'Generated an empty bundle'
+			});
+		}
 
-							timeEnd('sourcemap', 3);
-						}
+		this.hasDynamicImport = !!renderOptions.importMechanism;
+
+		this.renderedDeclarations = {
+			dependencies: this.getChunkDependencyDeclarations(options),
+			exports: this.getChunkExportDeclarations()
+		};
+	}
+
+	generateNamePreserveModules(preserveModulesRelativeDir: string) {
+		return (this.id = normalize(relative(preserveModulesRelativeDir, this.entryModule.id)));
+	}
 
-						if (code[code.length - 1] !== '\n') code += '\n';
-						return { code, map };
+	generateName(pattern: string, addons: Addons, existingNames?: { [name: string]: boolean }) {
+		// replace any chunk replacements
+		let outName = pattern.replace(/\[(hash|alias)\]/g, type => {
+			switch (type) {
+				case '[hash]':
+					return this.computeFullHash(addons);
+				case '[alias]':
+					if (!this.isEntryModuleFacade) {
+						return 'chunk';
+					} else if (this.entryModule.alias) {
+						return this.entryModule.alias;
+					} else {
+						let alias = basename(this.entryModule.id);
+						if (alias.endsWith('.js')) alias = alias.substr(0, alias.length - 3);
+						else if (alias.endsWith('.mjs')) alias = alias.substr(0, alias.length - 4);
+						return alias;
 					}
-				);
-			});
+			}
+		});
+
+		if (existingNames) {
+			if (!existingNames[outName]) {
+				existingNames[outName] = true;
+			} else {
+				let ext = extname(outName);
+				if (ext === '.js' || ext === '.mjs')
+					outName = outName.substr(0, outName.length - ext.length);
+				else ext = '';
```

Comment:
Again, I'm -1 to supporting other extensions. There are only two supported extensions for the JS language.

If a user really needed this, I'd rather allow registering custom js extensions.

Reply:
I am ok with leaving it this way but then I would prefer if you put the extensions into a constant for the three different places using them to make it easier to find them later.

Then (in another PR), we could add an option e.g. `extensions:` to register a set of extensions to replace this constant that
* will be considered extensions of JavaScript files when generating names
* can be omitted when importing files (and finally make everyone happy)

The default could then be `['.js' , '.mjs']`. Though this would diverge from our previous stance on this. But since we have a preference for certain extensions now anyway, maybe it is time.

Reply:
Included.